### DISCUSSION TITLE
Are there any restrictions on using gurobi in yalmip?

### DISCUSSION THREAD

**Discussion by ellensonny** [yalmip/YALMIP#1408] [General discussion] [answered]
hello Prof,
This is the part of my code that uses the gurobi solver：
![image](https://github.com/yalmip/YALMIP/assets/130203697/66dd55bc-6d98-4921-80a9-cbeead186589)
However, there is a problem here：
Warning: Solver not applicable (gurobi does not support quadratic semidefinite constraints).
Gurobi's Q&A staff told me: 1，gurobi itself can handle this problem
2，python supports all the functions of gurobi, but yalmip does not support the non-linear functions of gurobi.
Is there any way to use the full functionality of gurobi in yalmip without restrictions?

thanks for your reply very much!!!

**Comment by johanlofberg** [comment]
Gurobi does not support semidefinite programming.

Most likely you've defined a symmetric matrix and by mistake constrained it to be positive semidefinite instead of elementwise non-negative https://yalmip.github.io/tutorial/basics/. You see that if you display the object con

**Reply by ellensonny** [reply]
thank you for your reply！Actually, I purposely define a symmetric matrix.However, there must be something wrong somewhere, I will check it carefully later.
What really confuses me is that It seems that using gurobi in yalmip cannot handle non-linear problems. But this is not the case. I often encounter situations like "Warning: Solver not applicable (gurobi does not support ....constraints)" when programming.
Is it because gurobi handles linear problems by default in yalmip? Can set gurobi as a global solver to  solve this problem?

**Reply by johanlofberg** [reply]
Nonlinear (as in sin/exp etc) is a very recent addition to gurobi so support for that simply hasn't been added yet. It will though, as gurobi appears to use almost exactly YALMIPs internal format, and is very limited, so it will be easy to add. 

If you create a symmetric matrix, and want it to have non-negative elements, you will have to write that as X(:)>=0 or triu(X)>=0 as explained in the basic introduction

**Reply by johanlofberg** [reply]
Any particular function you use? I did a test and will basically be able to commit a working model directly

**Reply by ellensonny** [reply]
sorry to reply you so late.
Initially, my objective function looked like this:
![image](https://github.com/yalmip/YALMIP/assets/130203697/c5639b3d-2d75-42b6-a78d-3296c90b0dfe)

Except that G and B are constant matrices，other variables are as follows:
x(:,:,3)=sparse(row,col,ones(1,66),24,24).*binvar(24,24);
x(:,:,2)=sparse(row,col,ones(1,66),24,24).*binvar(24,24);
x(:,:,1)=sparse(row,col,ones(1,66),24,24).*binvar(24,24);
V=sdpvar(3,24,'full');
theta=sdpvar(3,24,'full');
 however, I  encountered situations like "Warning: Solver not applicable (gurobi does not support ....constraints)" when programming.

Then I did a test with the following simple model：
Using gurobi:
![image](https://github.com/yalmip/YALMIP/assets/130203697/366cc437-e449-4514-b9a1-3beada3632f7)
it works  without using gurobi:
![image](https://github.com/yalmip/YALMIP/assets/130203697/71cedbee-54f5-4787-a111-9c40fabdcc53)

So, calling gurobi in yalmip, why does this happen?

**Reply by johanlofberg** [reply]
As I said, gurobi is historically a MISOCP solver, i.e. not intended for general nonlinear programming, and requires much more thought to see when things can be cast as SOCP stuff. You cannot just throw arbitrary nonlinear stuff at it as a general local nonlinear solver such as fmincon. Since a couple of generations back it was extended to global nonconvex QCQP which YALMIP supports, and in the most recent version they added support for a restricted set of nonlinear operators, which YALMIP does not interface yet.

Cubic can be handled easily though, simply introduce a new variable z, and then add then constraint z == x1*x1 and then use f*z, as that will lead to a nonconvex QCQP

**Reply by ellensonny** [reply]
Yeah, I heard about that they added support for a restricted set of nonlinear operators.Since YALMIP does not interface yet,when I was describing this mathematical model, gurobi had no way to solve it.
![image](https://github.com/yalmip/YALMIP/assets/130203697/709639c3-db41-4820-ac6d-40a449ba06d0)
Would you give me some suggestions to solve this problem correctly without using gurobi or other methods?

**Reply by johanlofberg** [reply]
Initial fixes have been pushed to develop branch to try

**Reply by johanlofberg** [reply, marked as answer]
you're still going to have to massage the model though

the cubic expression x*V^2 should be written as Q^2 where you want Q=x*V which you most likely do best by logic linearization, i.e implies(x,Q==V) + implies(1-x,Q==0)

in the same way, x*V1*V2*cos(a) you write as Q*cos(a) where Q is x*Q1*V2 which you do by Q==Q2*V2 where Q2 is the logic linearization of x*V1

**Reply by ellensonny** [reply]
thanks a lot ! I will have a try.
